docs: Add guide for cluster with Docker Hub (#14)
sse-secure-systems/connaisseur: examples/eks/README.md

## Requirements

We assume you have a running EKS cluster already. Since this tutorial works with the CLI, we furthermore assume you have the `aws`, `docker`, `eksctl`, `git`, `kubectl`, `make` binaries installed and usable, i.e. having run `aws configure`, `aws eks update-kubeconfig` and `docker login`.

> This tutorial was tested on a machine running Ubuntu 20.04 with a EKS cluster based on EC2 instances.

## Starting and testing a deployment of Connaisseur

### 1. Set up environment

First off, we'll export the variables used in this tutorial to allow you to immediately use your own names/URLs instead of relying on our default names. Below, substitute your own values as appropriate:

```bash
REGISTRY_URL=docker.io
REPOSITORY_NAME=testingconny

> IMAGE_NAME=testserver

Reviewer: I would recommend using `testimage` as default value. `testserver` felt confusing to me at first sight.

IMAGE_PATH=$(echo ${REGISTRY_URL}/${REPOSITORY_NAME}/${IMAGE_NAME})

NOTARY_URL=notary.docker.io
NOTARY_USER=<YOUR-NOTARY-USERNAME/DOCKER-HUB-ID>
NOTARY_PASSWORD=<YOUR-NOTARY-PASSWORD>

git clone https://github.com/sse-secure-systems/connaisseur.git
cd connaisseur
```

> ### (Optional) 2. Prepare test images

Reviewer: I see you put this here, since we need the DCT root key later. Technically, it interrupts the flow and would be much better placed at the end before deploying the image. Might enforce creation of a DCT key at this point

Below is a list of commands to programmatically do the changes to the `value.yaml` file in the `helm` folder. The openssl command will take a password as input, so you won't be able to copy+paste past it. However, you can also do these changes manually if you want. They are comprised of setting the notary server, notary credentials and the root publiy key for verification:

```bash
sed -i "s/host: notary.docker.io/host: ${NOTARY_URL}/" helm/values.yaml
sed -i "s/user: notaryuser/user: ${NOTARY_USER}/" helm/values.yaml
sed -i "s/password: Password123/password: ${NOTARY_PASSWORD}/" helm/values.yaml

FIRST_LINE=$(grep -n rootPubKey: helm/values.yaml | sed s/:.*//)
LAST_LINE=$(grep -nF -- '-----END PUBLIC KEY-----' helm/values.yaml | sed s/:.*//)
sed -i $((${FIRST_LINE} + 1)),${LAST_LINE}d helm/values.yaml

cd ~/.docker/trust/private
sed '/^role:\sroot$/d' $(grep -iRl "role: root" .) > root-priv.key
openssl ec -in root-priv.key -pubout -out root-pub.pem
sed -i "${FIRST_LINE}s/.*/  rootPubKey: |\n    $(sed ':a;N;$!ba;s/\n/\\n    /g' root-pub.pem)/" ${OLDPWD}/helm/values.yaml
rm root-priv.key root-pub.pem
cd -

> ```

Reviewer: One could set Conny to detection mode which would make sure to not break anything, but might make seeing the outcome of using Conny less explicit (either warning or nothing).

Author: I wouldn't do that (by default). But since it is a nice feature, I would be inclined to mention it, now that you mention it (hehe). Like "hey if y'all wanna use non-blocking mode, you could also run "this command here", which is nice for checking at first but just doesnt give you as much sec, so y'all probably dont wanna do that in prod long-term.
That fine by you?

Reviewer: Would go for smth like that. Like "this could be useful for testing or rolling out", but wont block the ima just give a log message/warning

> ```

Reviewer: Might make a comment that this great for testing conny on a test cluster with test creds. For production use, one should look into the other options to manage the notary credentials and consider the non-strict detection mode. Might also resolve the comment above.

Author: Again, I would be hesitant to recommend non-strict for more than testing. Can add a comment around cred handling though. Do also have a different option to putting the key in this conf file, like deploying a secret to the namespace (haven't really gone deep into conny)? Maybe also @phbelitz can give a bit of guidance here for how to best manage these

> The `values.yaml` file in the `helm` folder is _the_ config file for Connaisseur. If you have some time to spare, have a look :)

### 4. Deploy Connaiseur

> After having completed the configuration, deploying Connaisseur is easy. Run
> 
> ```bash
> make install
> ```
> 
> after setting your Kubernetes context to the correct cluster. Your output should look something like

Reviewer: ```suggestion
Deploying Connaisseur is easy once the configuration is completed. Set your Kubernetes context to the correct cluster and run:

```bash
make install
```

Your output should look something like
```

```bash
service/sample created
Error from server: error when creating "STDIN": admission webhook "connaisseur-svc.connaisseur.svc" denied the request: could not find signed digest for image "docker.io/testingconny/testserver:unsigned" in trust data.
```

> Note that the service was indeed created, because it does not reference an image and is thus not blocked by Connaisseur. You can clean it up by calling `kubectl delete service sample`.

Finally, make sure that Connaisseur will deploy the signed image and isn't just rejecting all images:

```bash
kubectl create -f - << EOF
apiVersion: apps/v1
kind: Deployment
metadata:
  name: sample-deployment
  labels:
    app: sample
spec:
  selector:
    matchLabels:
      app: sample
  replicas: 1
  template:
    metadata:
      labels:
        app: sample
    spec:
      containers:
      - name: sample
        image: ${IMAGE_PATH}:signed
        imagePullPolicy: Always
        ports:
        - containerPort: 5000
---
apiVersion: v1
kind: Service
metadata:
  name: sample
  labels:
    app: sample
spec:
  ports:
  - name: port
    port: 8080
    targetPort: 5000
  selector:
    app: sample
  type: LoadBalancer
EOF

> ```

Reviewer: this and the following could technically also be example files, right?

Author: technically yes, but then you do need envsubst to substitute the IMAGE_PATH variable and envsubst is actually not installed by default and iirc needs quite a few dependencies. All the other tooling, git, make, kubectl, sed, grep, ... are either preinstalled on most systems or can be reasonably suspected to be on the machine of whoever tries the tutorial. Envsubst isnt, so I wouldnt do it

Author: Also tbh I like the fact that we show what's being run on k8s, instead of `kubecl apply -f thisfilewhichmayrunarbitraryprivilegeddeployments.yaml`. For the Dockerfiles it's not that important, since it will by default be run in the very deployment defined in this piece of code and thus can be considered somewhat sandboxed
